packages/medusa/src/services/payment-collection.ts

```diff
@@ -275,12 +273,114 @@ export default class PaymentCollectionService extends TransactionBaseService {
       }
 
       if (payCol.payment_sessions?.length) {
-        const removeIds: string[] = payCol.payment_sessions
-          .map((sess) => sess.id)
-          .filter((id) => !selectedSessionIds.includes(id))
+        const removeSessions: PaymentSession[] = payCol.payment_sessions.filter(
+          ({ id }) => !selectedSessionIds.includes(id)
+        )
+
+        if (removeSessions.length) {
+          await paymentCollectionRepository.deleteMultiple(
+            removeSessions.map((sess) => sess.id)
+          )
 
-        if (removeIds.length) {
-          await paymentCollectionRepository.deleteMultiple(removeIds)
+          Promise.all(
+            removeSessions.map(async (sess) =>
+              this.paymentProviderService_
+                .withTransaction(manager)
+                .deleteSessionNew(sess)
+            )
+          ).catch(() => void 0)
+        }
+      }
+
+      payCol.payment_sessions = paymentSessions
+
+      return await paymentCollectionRepository.save(payCol)
+    })
+  }
+
+  async setPaymentSession(
+    paymentCollectionId: string,
+    sessionInput: PaymentCollectionsSessionsInput,
+    customer_id: string
+  ): Promise<PaymentCollection> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      const payCol = await this.retrieve(paymentCollectionId, {
+        relations: ["region", "region.payment_providers", "payment_sessions"],
+      })
+
+      if (payCol.status !== PaymentCollectionStatus.NOT_PAID) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Cannot set payment sessions for a payment collection with status ${payCol.status}`
+        )
+      }
+
+      const hasProvider = payCol?.region?.payment_providers
+        .map((p) => p.id)
+        .includes(sessionInput.provider_id)
+
+      if (!hasProvider) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "Payment provider not found"
+        )
+      }
+
+      const customer = await this.customerService_
+        .withTransaction(manager)
+        .retrieve(customer_id, {
+          select: ["id", "email", "metadata"],
+        })
+
+      const paymentSessions: PaymentSession[] = []
+      const inputData: PaymentProviderDataInput = {
+        resource_id: payCol.id,
+        currency_code: payCol.currency_code,
+        amount: payCol.amount,
+        provider_id: sessionInput.provider_id,
+        customer,
+        metadata: {
+          resource_id: payCol.id,
```

Comment:
thought: is this necessary?  since there already is a resource id 6 lines above?

Reply:
Will check that. Pretty sure that is being used by a payment plugin, but maybe that can be adapted there.

Reply:
nice catch.. Thanks.
I mixed a few concepts. This metadata is not needed.